flang/lib/Optimizer/Builder/TemporaryStorage.cpp

```diff
@@ -258,13 +258,19 @@ void fir::factory::AnyVariableStack::pushValue(mlir::Location loc,
                                                fir::FirOpBuilder &builder,
                                                mlir::Value variable) {
   hlfir::Entity entity{variable};
-  mlir::Type storageElementType =
-      hlfir::getFortranElementType(retValueBox.getType());
-  auto [box, maybeCleanUp] =
-      hlfir::convertToBox(loc, builder, entity, storageElementType);
-  fir::runtime::genPushDescriptor(loc, builder, opaquePtr, fir::getBase(box));
-  if (maybeCleanUp)
-    (*maybeCleanUp)();
+  if (mlir::isa<fir::BaseBoxType>(entity.getType())) {
```

Comment:
Would using `hlfir::genVariableBox` here instead of convertToBox solve the issue? (no need to provide a forceBoxType also slove the issue).

If so, that would simplify the code.